### PR TITLE
Add FlexGroup component

### PR DESCRIPTION
**Pull request by harrisonturton**
Warning, huge PR. There are 1500+ lines of tests 🤧 I'd split this up if I wasn't working by myself.

This PR is to implement two new components:

* `FlexGroup`
* `Flex`

Which implement a similar layout algorithm as flexbox on the web and flex widgets in Flutter.

It also makes a series of related changes to other modules, like `Vector2`, and implements a new `test_util` workspace crate. I'm hoping the tests set some precedent for how I can begin testing other widgets.

### CODE REVIEW

**Comment by harrisonturton** [comment]
Hmm... I'm having some trouble making this API easy to use.

Currently all children of `FlexGroup` must be wrapped in a `Flex` enum:

```rust
let widgets = FlexGroup {
    children: vec![
        Flex::Fixed {
            child: Box::new(/* fixed child */),
        },
        Flex::Flexible {
            flex: 1.0,
            child: Box::new(/* flexible child */),
        },
    ],
    ..Default::default()
};
```

This is because `FlexGroup` needs a way to tell if a child is flexible, and if so, what flex factor it has. But I don't love this API. It's not very ergonomic. Being forced to wrap everything in a `Flex` enum is annoying.

I'd love to simplify this, and only wrap the child in `Flex` type if we need to provide the flex factor explicitly:

```rust
let widgets = FlexGroup {
    children: vec![
        Box::new(/* fixed child */),
        Flex {
            flex: 1.0,
            child: Box::new(/* flexible child */),
        },
    ],
    ..Default::default()
};
```

In other words, I'd love to be able to define fixed-sized children directly.

But I can't seem to get it to work! I'm sure there's a possible solution with traits/supertraits/generics. Time to look for a workaround!

**Comment by harrisonturton** [comment]
I've figured out a workaround for the flex factor issue! Metatraits to the rescue.

I can do something like this:

```rust
trait FlexLayout {
    fn flex_factor(&self) -> f32;
}

impl<T> FlexLayout for T where T: Layout {
    fn flex_factor(&self) -> f32 {
        0.0
    }
}

struct Flex {
    flex_factor: f32,
    child: Box<dyn Layout>
}

impl FlexLayout for Flex {
    fn flex_factor(&self) -> f32 {
        self.flex_factor
    }
}

struct FlexGroup {
    children: Vec<Box<dyn FlexLayout>>
}
```

The magic line of `impl<T> FlexLayout for T where T: Layout` implements the `FlexLayout` trait on all instances of the `Layout` trait! Which means we can then use `Box<dyn Layout>` wherever we can use `Box<dyn FlexLayout>`.

I think I'll do this in another PR, through. This one is getting pretty big, and I want to merge it into main.

**Comment by harrisonturton** [comment]
> Hmm... I'm having some trouble making this API easy to use.

Fixed in #2
